### PR TITLE
feat: supporting consent mode for GARL

### PR DESCRIPTION
**Pull request by shrouti1507**
## What are the changes introduced in this PR?

We are adding support for consent field in Google Adwords remarketing list. We are expecting the fields will be populated from VDM and have named them as `adUserData_consent` , `adPersonalization_consent`, to avoid any ambiguity internally and using these values send provided consent for each user level.

As this is an audiencelist event type destination, where multiple user info blocks are expected, so we needed to group the users based on similar consent modes and send to GARL

## What is the related Linear task?

Resolves INT-1348

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [x] All related docs linked with the PR?

- [ ] All changes manually tested?

- [x] Any documentation changes needed with this change?

- [x] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3120/test-report.html

File: src/v0/destinations/google_adwords_remarketing_lists/utils.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        adUserData: validateAndTransformConsent(detail.adUserData_consent),
+        adPersonalization: validateAndTransformConsent(detail.adPersonalization_consent),
+      }))
+      .groupBy((detail) => `${detail.adUserData}-${detail.adPersonalization}`)
+      .map((group, key) => {
+        const consentParts = key.split('-');
+        // Adjust the action key based on the original action (add or remove)
+        //           const actionKey = action === 'add' ? 'create' : action;
+        return {
+          consent: { adUserData: consentParts[0], adPersonalization: consentParts[1] },
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/v0/destinations/google_adwords_remarketing_lists/utils.js:86]: Though it's internal, let's use proper naming.

File: src/v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 2]
```diff
@@ -209,17 +219,30 @@ const processEvent = async (metadata, message, destination) => {
...
   }
   if (message.type.toLowerCase() === 'audiencelist') {
-    const createdPayload = createPayload(message, destination);
+    const userDataGroupedByConsentLevel = groupUserDataBasedOnConsentLevel(message);
+    userDataGroupedByConsentLevel.forEach((userDataBlock) => {
+      createdPayload = { ...createPayload(message, userDataBlock, destination) };
+      payloadArray.push(createdPayload);
+    });
+
+    const sortedPayloadArray = lodash.sortBy(payloadArray, (item) => !item.remove);
```
Comment (saikumarrs) [inline src/v0/destinations/google_adwords_remarketing_lists/transform.js:228]: Why do we want to sort the array?

File: src/v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 3]
```diff
@@ -209,17 +219,30 @@ const processEvent = async (metadata, message, destination) => {
...
 
-    Object.values(createdPayload).forEach((data) => {
-      const consentObj = populateConsentForGoogleDestinations(message.properties);
-      response.push(responseBuilder(metadata, data, destination, message, consentObj));
+    sortedPayloadArray.forEach((singlePayload) => {
+      if (singlePayload && Object.keys(singlePayload).length <= 1) {
+        throw new InstrumentationError(
+          "Neither 'add' nor 'remove' property is present inside 'listData' or there are no attributes inside 'add' or 'remove' properties matching with the schema fields. Aborting message.",
+        );
+      }
```
Comment (saikumarrs) [inline src/v0/destinations/google_adwords_remarketing_lists/transform.js:241]: Is this scenario feasible?

File: src/v0/destinations/google_adwords_remarketing_lists/utils.test.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
+const { groupUserDataBasedOnConsentLevel } = require('./utils');
+
+describe('groupUserDataBasedOnConsentLevel', () => {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/v0/destinations/google_adwords_remarketing_lists/utils.test.js:3]: Also, add a scenario where a consent group contains both add and remove action data.

File: src/v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 5]
```diff
@@ -209,17 +219,30 @@ const processEvent = async (metadata, message, destination) => {
     throw new InstrumentationError('listData is not present inside properties. Aborting message.');
   }
   if (message.type.toLowerCase() === 'audiencelist') {
-    const createdPayload = createPayload(message, destination);
+    const userDataGroupedByConsentLevel = groupUserDataBasedOnConsentLevel(message);
+    userDataGroupedByConsentLevel.forEach((userDataBlock) => {
+      createdPayload = { ...createPayload(message, userDataBlock, destination) };
```
Comment (saikumarrs) [inline src/v0/destinations/google_adwords_remarketing_lists/transform.js:224]: I believe there is no need to use any object spread operator here.

File: src/v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 6]
```diff
@@ -209,17 +219,30 @@ const processEvent = async (metadata, message, destination) => {
...
-    Object.values(createdPayload).forEach((data) => {
-      const consentObj = populateConsentForGoogleDestinations(message.properties);
-      response.push(responseBuilder(metadata, data, destination, message, consentObj));
+    sortedPayloadArray.forEach((singlePayload) => {
+      if (singlePayload && Object.keys(singlePayload).length <= 1) {
+        throw new InstrumentationError(
+          "Neither 'add' nor 'remove' property is present inside 'listData' or there are no attributes inside 'add' or 'remove' properties matching with the schema fields. Aborting message.",
+        );
+      }
+      const finalDataObject = singlePayload.remove || singlePayload.create;
```
Comment (saikumarrs) [inline src/v0/destinations/google_adwords_remarketing_lists/transform.js:242]: We need to build the response for both operations, right?

File: src/v0/destinations/google_adwords_remarketing_lists/utils.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+ *     remove: [
+ *       {
+ *         email: "user3@example.com"
+ *       }
+ *     ]
+ *   }
+ * ]
+ *
+ */
+function groupUserDataBasedOnConsentLevel(input) {
```
Comment (saikumarrs) [inline src/v0/destinations/google_adwords_remarketing_lists/utils.js:69]: Can we move this utility outside this destination and make it more generic?

Take the consent field names and supported actions as arguments.

File: test/integrations/destinations/google_adwords_remarketing_lists/router/data.ts [hunk 8]
```diff
@@ -228,7 +228,14 @@ export const data = [
                     'Content-Type': 'application/json',
                     'developer-token': 'ijkl9101',
                   },
-                  params: { listId: '7090784486', customerId: '7693729833', consent: {} },
+                  params: {
```
Comment (saikumarrs) [inline test/integrations/destinations/google_adwords_remarketing_lists/router/data.ts:231]: We need to add new test cases for the scenarios where consent data is specified.

File: test/integrations/destinations/google_adwords_remarketing_lists/processor/data.ts [hunk 9]
```diff
@@ -79,7 +79,10 @@ export const data = [
               params: {
```
Comment (saikumarrs) [inline test/integrations/destinations/google_adwords_remarketing_lists/processor/data.ts:79]: We need to add new test cases for the scenarios where consent data is specified.

File: test/integrations/destinations/google_adwords_remarketing_lists/processor/data.ts [hunk 10]
```diff
@@ -2829,14 +2841,17 @@ export const data = [
...
+                  adUserData: 'UNSPECIFIED',
+                },
               },
               body: {
                 JSON: {
                   enablePartialFailure: true,
                   operations: [
                     {
-                      create: {
+                      remove: {
```
Comment (saikumarrs) [inline test/integrations/destinations/google_adwords_remarketing_lists/processor/data.ts:2854]: The operation change here. Is this expected?

**Review by saikumarrs — CHANGES_REQUESTED** [review]

**Comment by shrouti1507** [comment]
raised https://github.com/rudderlabs/rudder-transformer/pull/3143 after discussion. Closing this

**Pull request closed without merging**